cgen error when calling function with receiver that returns generic result

OS: windows, Microsoft Windows 11 Home v22000 64-bit
Processor: 20 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i7-12700
CC version: Error: exec failed (CreateProcess) with code 2: 系統找不到指定的檔案。
 cmd: cc --version

getwd: C:\Users\chaos\projects\v_test
vmodules: C:\Users\chaos\.vmodules
vroot: C:\Users\chaos\projects\v
vexe: C:\Users\chaos\projects\v\v.exe
vexe mtime: 2022-08-19 20:52:38
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.0 22a79cf.fa44744

Git version: git version 2.35.1.windows.2
Git vroot status: weekly.2022.33-28-gfa447443 (3 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 125e21e2

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
`v run generic_result.v`

generic_result.v
```v
struct Dummy {}

fn (dummy &Dummy) res<T>() !T {
	$if T is int {
		return 1
	} $else $if T is f32 {
		return dummy.res<int>()! + 1
	} $else {
		return error('exhausted')
	}
}

d := Dummy{}

println(d.res<int>()!)
println(d.res<f32>()!)
```

**What did you expect to see?**
```
1
2.
```
 
**What did you see instead?**
```
==================
C:/Users/chaos/AppData/Local/Temp/v_0/generic_result.7277512356653479659.tmp.c:7400: warning: cast between pointer and integer of different size
C:/Users/chaos/AppData/Local/Temp/v_0/generic_result.7277512356653479659.tmp.c:12312: error: '{' expected (got ";")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.
```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -307,7 +307,8 @@ fn (mut c Checker) fn_decl(mut node ast.FnDecl) {
 		}
 
 		if node.return_type != ast.void_type_idx
-			&& node.return_type.clear_flag(.optional) != ast.void_type_idx {
+			&& node.return_type.clear_flag(.optional) != ast.void_type_idx
+			&& node.return_type.clear_flag(.result) != ast.void_type_idx {
 			c.error('test functions should either return nothing at all, or be marked to return `?`',
 				node.pos)
 		}
@@ -1318,6 +1319,7 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 		}
 		node.is_noreturn = method.is_noreturn
 		node.is_ctor_new = method.is_ctor_new
+		node.return_type = method.return_type
 		if !method.is_pub && !c.pref.is_test && method.mod != c.mod {
 			// If a private method is called outside of the module
 			// its receiver type is defined in, show an error.
@@ -1550,21 +1552,16 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 			c.table.register_fn_concrete_types(method.fkey(), concrete_types)
 		}
 
-		// resolve return generics struct to concrete type
-		if method.generic_names.len > 0 && method.return_type.has_flag(.generic)
-			&& !isnil(c.table.cur_fn) && c.table.cur_fn.generic_names.len == 0 {
-			node.return_type = c.table.unwrap_generic_type(method.return_type, method.generic_names,
-				concrete_types)
-		} else {
-			node.return_type = method.return_type
-		}
-		if node.concrete_types.len > 0 && method.return_type != 0 && !isnil(c.table.cur_fn)
-			&& c.table.cur_fn.generic_names.len == 0 {
+		if node.concrete_types.len > 0 && node.concrete_types.all(!it.has_flag(.generic))
+			&& method.return_type.has_flag(.generic) && method.generic_names.len > 0
+			&& method.generic_names.len == node.concrete_types.len {
 			if typ := c.table.resolve_generic_to_concrete(method.return_type, method.generic_names,
 				concrete_types)
 			{
 				node.return_type = typ
-				return typ
+			} else {
+				node.return_type = c.table.unwrap_generic_type(method.return_type, method.generic_names,
+					concrete_types)
 			}
 		}
 		if node.concrete_types.len > 0 && method.generic_names.len == 0 {
@@ -1578,9 +1575,8 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 					}
 				}
 			}
-			return node.return_type
 		}
-		return method.return_type
+		return node.return_type
 	}
 	// TODO: str methods
 	if method_name == 'str' {

--- a/vlib/v/tests/generics_method_with_nested_generic_method_test.v
+++ b/vlib/v/tests/generics_method_with_nested_generic_method_test.v
@@ -1,0 +1,23 @@
+struct Dummy {}
+
+fn (dummy &Dummy) res<T>() !T {
+	$if T is int {
+		return 1
+	} $else $if T is f32 {
+		return dummy.res<int>()! + 1
+	} $else {
+		return error('exhausted')
+	}
+}
+
+fn test_generics_method_with_nested_generic_method() ! {
+	d := Dummy{}
+
+	println(d.res<int>()!)
+	ret1 := d.res<int>()!
+	assert ret1 == 1
+
+	println(d.res<f32>()!)
+	ret2 := d.res<f32>()!
+	assert ret2 == 2.0
+}

